// arbundles/build/node/esm/src/nodeUtils.js
import { createPublicKey } from "crypto";
import { default as nodeDriver } from "arweave/node/lib/crypto/node-driver.js";
// import CryptoInterface from "arweave/node/lib/crypto/crypto-interface.js";
export { stringToBuffer, concatBuffers } from "arweave/node/lib/utils.js";
export { default as Transaction } from "arweave/node/lib/transaction.js";
export { deepHash } from "./deepHash.js";
export { default as Arweave } from "arweave/node/index.js";
// hack as ESM won't unpack .default CJS imports, so we do so dynamically
// eslint-disable-next-line @typescript-eslint/dot-notation
const driver = nodeDriver["default"] ? nodeDriver["default"] : nodeDriver;
export class CryptoDriver extends driver {
    getPublicKey(jwk) {
        return createPublicKey({
            key: this.jwkToPem(jwk),
            type: "pkcs1",
            format: "pem",
        })
            .export({
            format: "pem",
            type: "pkcs1",
        })
            .toString();
    }
}
let driverInstance;
export function getCryptoDriver() {
    return (driverInstance ??= new CryptoDriver());
}
