// @bundlr-network/client/build/cjs/common/api.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const bundlr_1 = __importDefault(require("./bundlr"));
// TODO: overhaul this
class Api {
    constructor(config) {
        this.METHOD_GET = "GET";
        this.METHOD_POST = "POST";
        this.cookieMap = new Map();
        this.applyConfig(config);
    }
    applyConfig(config) {
        this.config = this.mergeDefaults(config);
    }
    getConfig() {
        return this.config;
    }
    requestInterceptor(request) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const cookies = this.cookieMap.get(new URL((_a = request.baseURL) !== null && _a !== void 0 ? _a : "").host);
            if (cookies)
                request.headers.cookie = cookies;
            return request;
        });
    }
    responseInterceptor(response) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const setCookie = (_a = response.headers) === null || _a === void 0 ? void 0 : _a["set-cookie"];
            if (setCookie)
                this.cookieMap.set(response.request.host, setCookie);
            return response;
        });
    }
    mergeDefaults(config) {
        var _a, _b, _c, _d, _e, _f;
        const protocol = (_a = config.protocol) !== null && _a !== void 0 ? _a : "http";
        const port = (_b = config.port) !== null && _b !== void 0 ? _b : (protocol === "https" ? 443 : 80);
        return {
            host: (_c = config.host) !== null && _c !== void 0 ? _c : "127.0.0.1",
            protocol,
            port,
            timeout: (_d = config.timeout) !== null && _d !== void 0 ? _d : 20000,
            logging: (_e = config.logging) !== null && _e !== void 0 ? _e : false,
            logger: (_f = config.logger) !== null && _f !== void 0 ? _f : console.log,
            headers: Object.assign(Object.assign({}, config.headers), { "x-bundlr-js-sdk-version": bundlr_1.default.VERSION }),
            withCredentials: true,
        };
    }
    get(endpoint, config) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                return yield this.request().get(endpoint, Object.assign(Object.assign({}, config), { headers: Object.assign(Object.assign({}, config === null || config === void 0 ? void 0 : config.headers), (_a = this.config) === null || _a === void 0 ? void 0 : _a.headers) }));
            }
            catch (error) {
                if ((_b = error.response) === null || _b === void 0 ? void 0 : _b.status) {
                    return error.response;
                }
                throw error;
            }
        });
    }
    post(endpoint, body, config) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                return yield this.request().post(endpoint, body, Object.assign(Object.assign({}, config), { headers: Object.assign(Object.assign({}, config === null || config === void 0 ? void 0 : config.headers), (_a = this.config) === null || _a === void 0 ? void 0 : _a.headers) }));
            }
            catch (error) {
                if ((_b = error.response) === null || _b === void 0 ? void 0 : _b.status) {
                    return error.response;
                }
                throw error;
            }
        });
    }
    request() {
        if (this.instance)
            return this.instance;
        const instance = axios_1.default.create({
            baseURL: `${this.config.protocol}://${this.config.host}:${this.config.port}`,
            timeout: this.config.timeout,
            maxContentLength: 1024 * 1024 * 512,
            headers: this.config.headers,
            withCredentials: this.config.withCredentials,
        });
        if (this.config.withCredentials) {
            instance.interceptors.request.use(this.requestInterceptor.bind(this));
            instance.interceptors.response.use(this.responseInterceptor.bind(this));
        }
        if (this.config.logging) {
            instance.interceptors.request.use((request) => {
                this.config.logger(`Requesting: ${request.baseURL}/${request.url}`);
                return request;
            });
            instance.interceptors.response.use((response) => {
                this.config.logger(`Response:   ${response.config.url} - ${response.status}`);
                return response;
            });
        }
        return (this.instance = instance);
    }
}
exports.default = Api;
